File: internal/cli/report_status.go
package cli

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"

	"github.com/andrewcohen/awp/internal/state"
	"github.com/andrewcohen/awp/internal/workspace"
)

// validReportStates is the closed set of states agents may report.
var validReportStates = map[string]struct{}{
	"working": {},
	"idle":    {},
	"waiting": {},
	"exited":  {},
}

// runReportStatus is the entry point for `awp internal report-status`.
//
// It is invoked by per-agent hooks/extensions installed globally via
// `awp init hooks`. The hook command resolves the workspace via
// $AWP_WORKSPACE (workspace name) and one of:
//   - $AWP_REPO_ROOT (preferred, absolute repo root path)
//   - $AWP_REPO      (project basename; ambiguous if multiple repos share it)
//
// When env vars are missing the command exits 0 silently so a misconfigured
// hook never breaks an agent turn.
//
// Optional flags capture the active prompt alongside the state transition:
//   - --prompt <text>     persist the literal text as the workspace's ActivePrompt.
//   - --prompt-stdin      read a Claude-style hook payload JSON from stdin and
//                         extract its top-level "prompt" field. Empty/missing
//                         is treated as "no prompt update" rather than an error.
func runReportStatus(args []string, out io.Writer) error {
	if isHelpArgSlice(args) {
		_, _ = fmt.Fprintln(out, "Usage: awp internal report-status --state <working|idle|waiting|exited> [--prompt <text>|--prompt-stdin]")
		return nil
	}
	state := ""
	prompt := ""
	promptStdin := false
	for i := 0; i < len(args); i++ {
		arg := args[i]
		switch {
		case arg == "--state":
			if i+1 >= len(args) {
				return fmt.Errorf("--state requires a value")
			}
			state = args[i+1]
			i++
		case strings.HasPrefix(arg, "--state="):
			state = strings.TrimPrefix(arg, "--state=")
		case arg == "--prompt":
			if i+1 >= len(args) {
				return fmt.Errorf("--prompt requires a value")
			}
			prompt = args[i+1]
			i++
		case strings.HasPrefix(arg, "--prompt="):
			prompt = strings.TrimPrefix(arg, "--prompt=")
		case arg == "--prompt-stdin":
			promptStdin = true
		default:
			return fmt.Errorf("unknown argument %q", arg)
		}
	}
	state = strings.ToLower(strings.TrimSpace(state))
	if state == "" {
		return errors.New("--state is required")
	}
	if _, ok := validReportStates[state]; !ok {
		return fmt.Errorf("invalid --state %q (want working|idle|waiting|exited)", state)
	}
	if promptStdin {
		// Best-effort: a malformed payload should never break the agent turn.
		// Silently drop errors and fall through with prompt="".
		if stdinPrompt, ok := readPromptFromStdin(reportStatusStdin()); ok {
			prompt = stdinPrompt
		}
	}
	prompt = strings.TrimSpace(prompt)

	workspaceName, repoName, repoRoot := resolveWorkspaceIdent()
	if workspaceName == "" {
		return nil
	}
	return writeWorkspaceStatus(workspaceName, repoName, repoRoot, state, prompt)
}

// readPromptFromStdin parses Claude's UserPromptSubmit hook payload and
// returns the "prompt" string. Returns (_, false) when stdin is empty, not
// JSON, or the field is missing — callers should treat that as "no prompt
// update", not an error.
func readPromptFromStdin(r io.Reader) (string, bool) {
	data, err := io.ReadAll(r)
	if err != nil || len(data) == 0 {
		return "", false
	}
	var payload struct {
		Prompt string `json:"prompt"`
	}
	if err := json.Unmarshal(data, &payload); err != nil {
		return "", false
	}
	if payload.Prompt == "" {
		return "", false
	}
	return payload.Prompt, true
}

// resolveWorkspaceIdent returns (AWP_WORKSPACE, AWP_REPO, AWP_REPO_ROOT) with
// a tmux fallback. When the process env is empty (e.g. the calling Claude/pi
// was launched before the tmux session env was injected), we ask tmux for the
// session-level values. This makes hooks robust against stale process
// environments.
func resolveWorkspaceIdent() (workspace, repo, repoRoot string) {
	workspace = strings.TrimSpace(os.Getenv("AWP_WORKSPACE"))
	repo = strings.TrimSpace(os.Getenv("AWP_REPO"))
	repoRoot = strings.TrimSpace(os.Getenv("AWP_REPO_ROOT"))
	if workspace != "" {
		return
	}
	if strings.TrimSpace(os.Getenv("TMUX")) == "" {
		return
	}
	if v := tmuxLocalEnv("AWP_WORKSPACE"); v != "" {
		workspace = v
	}
	if repo == "" {
		repo = tmuxLocalEnv("AWP_REPO")
	}
	if repoRoot == "" {
		repoRoot = tmuxLocalEnv("AWP_REPO_ROOT")
	}
	return
}

// tmuxLocalEnv reads a single session-level env var from the tmux server.
// We resolve the current session via display-message and pin show-environment
// with `-t` rather than relying on tmux's implicit "current session" — that
// inference depends on $TMUX_PANE being set and the pane still existing,
// which isn't always true for hook child processes.
//
// Returns empty on any error, when TMUX is unset, or when the var is unset
// or explicitly removed (`-KEY` form).
func tmuxLocalEnv(key string) string {
	if strings.TrimSpace(os.Getenv("TMUX")) == "" {
		return ""
	}
	sessionOut, err := exec.Command("tmux", "display-message", "-p", "#{session_name}").Output()
	if err != nil {
		return ""
	}
	session := strings.TrimSpace(string(sessionOut))
	if session == "" {
		return ""
	}
	out, err := exec.Command("tmux", "show-environment", "-t", session, key).Output()
	if err != nil {
		return ""
	}
	line := strings.TrimSpace(string(out))
	if line == "" || strings.HasPrefix(line, "-") {
		return ""
	}
	if idx := strings.IndexByte(line, '='); idx >= 0 {
		return strings.TrimSpace(line[idx+1:])
	}
	return ""
}

// writeWorkspaceStatus mutates Status (and optionally ActivePrompt) on the
// matching entry. It prefers repoRoot (absolute path) for an exact match;
// falls back to repoName (basename of each known repo root) when the root
// is unknown. It also flips Unread=true on transitions into "attention"
// states so the tmux badge surfaces the change.
//
// ActivePrompt lifecycle: a non-empty prompt argument overwrites the field
// (UserPromptSubmit / before_agent_start path). When the new status is
// "idle" or "exited" we clear ActivePrompt because the agent is no longer
// acting on that prompt. "working" and "waiting" leave it alone so the
// deck keeps showing the prompt while the agent is mid-task or pinging
// for attention.
func writeWorkspaceStatus(workspaceName, repoName, repoRoot, status, prompt string) error {
	store := stateStore()

	// Suppress the badge when the user is literally looking at this
	// workspace's session — same logic as the deck's auto-clear, applied
	// at write time so the tmux status bar stays accurate without waiting
	// for a deck refresh.
	viewing := sessionHasAttachedClient(repoName, workspaceName)
	apply := func(entries map[string]workspace.Entry) map[string]workspace.Entry {
		entry, ok := entries[workspaceName]
		if !ok {
			return entries
		}
		entry.Status = status
		switch {
		case prompt != "":
			entry.ActivePrompt = prompt
		case status == "idle" || status == "exited":
			entry.ActivePrompt = ""
		}
		if workspace.WantsAttention(status) {
			if viewing {
				entry.Unread = false
			} else {
				entry.Unread = true
			}
		}
		entries[workspaceName] = entry
		return entries
	}

	if repoRoot != "" {
		if u, ok := store.(updater); ok {
			return u.Update(repoRoot, apply)
		}
		entries, err := store.Load(repoRoot)
		if err != nil {
			return err
		}
		entries = apply(entries)
		return store.Save(repoRoot, entries)
	}

	if repoName == "" {
		return nil
	}
	all, err := store.LoadAll()
	if err != nil {
		return err
	}
	// Deterministic basename fallback: collect every repo whose basename
	// matches, sort for stability, then prefer one that actually has the
	// named workspace as an entry. If exactly one match has the entry, we
	// write to it. If multiple have it (basename collision across repos
	// that both happen to have a same-named workspace), we no-op rather
	// than silently route to an arbitrary pick — better to drop a status
	// than to badge the wrong workspace.
	var candidates []string
	for root := range all {
		if filepath.Base(root) == repoName {
			candidates = append(candidates, root)
		}
	}
	sort.Strings(candidates)
	var matches []string
	for _, root := range candidates {
		if _, ok := all[root][workspaceName]; ok {
			matches = append(matches, root)
		}
	}
	switch len(matches) {
	case 0:
		// Fall back to first basename match — preserves prior behavior
		// when there's no entry collision (most common case: one repo
		// shares the basename and the entry hasn't been created yet,
		// e.g. status arrives during workspace setup).
		if len(candidates) == 0 {
			return nil
		}
		root := candidates[0]
		if u, ok := store.(updater); ok {
			return u.Update(root, apply)
		}
		entries, err := store.Load(root)
		if err != nil {
			return err
		}
		entries = apply(entries)
		return store.Save(root, entries)
	case 1:
		root := matches[0]
		if u, ok := store.(updater); ok {
			return u.Update(root, apply)
		}
		entries, err := store.Load(root)
		if err != nil {
			return err
		}
		entries = apply(entries)
		return store.Save(root, entries)
	default:
		return nil
	}
}

type updater interface {
	Update(repoRoot string, fn func(map[string]workspace.Entry) map[string]workspace.Entry) error
}

// sessionHasAttachedClient reports whether at least one tmux client is
// currently attached to the workspace's session — i.e. the user is looking
// at it. Best-effort: any tmux/exec error returns false (we'd rather badge
// than silently miss).
func sessionHasAttachedClient(repoName, workspaceName string) bool {
	repoName = strings.TrimSpace(repoName)
	workspaceName = strings.TrimSpace(workspaceName)
	if repoName == "" || workspaceName == "" {
		return false
	}
	if strings.TrimSpace(os.Getenv("TMUX")) == "" {
		return false
	}
	session := DeckSessionName(repoName, workspaceName)
	out, err := exec.Command("tmux", "list-clients", "-t", session, "-F", "#{client_name}").Output()
	if err != nil {
		return false
	}
	return strings.TrimSpace(string(out)) != ""
}

// stateStore returns a JSONStore. Indirection exists so tests can swap it.
var stateStore = func() reportStatusStore { return state.NewJSONStore() }

// reportStatusStdin returns the reader used by --prompt-stdin. Indirection
// exists so tests can stub in a buffer without touching os.Stdin.
var reportStatusStdin = func() io.Reader { return os.Stdin }

type reportStatusStore interface {
	Load(repoRoot string) (map[string]workspace.Entry, error)
	LoadAll() (map[string]map[string]workspace.Entry, error)
	Save(repoRoot string, entries map[string]workspace.Entry) error
}

File: rust/crates/awp-agent/src/report_status.rs
//! `awp report-status` — an agent hook writes a workspace's status to the store.
//!
//! Ports the Go `internal/cli/report_status.go`, with the whole-file JSON
//! rewrite replaced by a partial, row-level SQLite `UPDATE`. Must-preserve
//! behaviors:
//!
//! - Closed set of reportable states (working/idle/waiting/exited).
//! - `--prompt` / `--prompt-stdin` (Claude payload `prompt` field) /
//!   `--waiting-when-tool` (PreToolUse `tool_name` → override to waiting).
//! - Resolve the workspace via `$AWP_WORKSPACE` + (`$AWP_REPO_ROOT` |
//!   `$AWP_REPO`); no-op silently when identity is missing or the row is
//!   unknown, so a misconfigured hook never breaks an agent turn.
//! - ActivePrompt lifecycle: non-empty prompt overwrites; idle/exited clears;
//!   working/waiting leave it. Unread: exited clears; waiting/idle set (unless
//!   viewing); working preserves.

use anyhow::{bail, Result};
use awp_core::{Status, WorkspaceId};
use awp_store::Store;
use std::io::Read;

/// Parsed `report-status` arguments.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct ReportArgs {
    pub state: String,
    pub prompt: Option<String>,
    pub prompt_stdin: bool,
    pub waiting_tools: Vec<String>,
}

/// Workspace identity from the environment (or tmux fallback, resolved by the
/// caller).
#[derive(Debug, Default, Clone)]
pub struct Ident {
    pub workspace: String,
    pub repo: String,
    pub repo_root: String,
}

impl Ident {
    /// Resolve from the process environment. Mirrors Go `resolveWorkspaceIdent`
    /// (minus the tmux `show-environment` fallback, which the deck's own env
    /// injection makes unnecessary in the common case).
    pub fn from_env() -> Self {
        let get = |k: &str| std::env::var(k).unwrap_or_default().trim().to_string();
        Ident {
            workspace: get("AWP_WORKSPACE"),
            repo: get("AWP_REPO"),
            repo_root: get("AWP_REPO_ROOT"),
        }
    }
}

/// Parse the CLI args. Unknown flags are an error (matching Go).
pub fn parse_args(args: &[String]) -> Result<ReportArgs> {
    let mut out = ReportArgs::default();
    let mut i = 0;
    while i < args.len() {
        let arg = args[i].as_str();
        match arg {
            "--state" => {
                i += 1;
                out.state = args
                    .get(i)
                    .cloned()
                    .ok_or_else(|| anyhow::anyhow!("--state requires a value"))?;
            }
            _ if arg.starts_with("--state=") => {
                out.state = arg.trim_start_matches("--state=").to_string();
            }
            "--prompt" => {
                i += 1;
                out.prompt = Some(
                    args.get(i)
                        .cloned()
                        .ok_or_else(|| anyhow::anyhow!("--prompt requires a value"))?,
                );
            }
            _ if arg.starts_with("--prompt=") => {
                out.prompt = Some(arg.trim_start_matches("--prompt=").to_string());
            }
            "--prompt-stdin" => out.prompt_stdin = true,
            "--waiting-when-tool" => {
                i += 1;
                let v = args
                    .get(i)
                    .ok_or_else(|| anyhow::anyhow!("--waiting-when-tool requires a value"))?;
                out.waiting_tools = parse_tool_list(v);
            }
            _ if arg.starts_with("--waiting-when-tool=") => {
                out.waiting_tools = parse_tool_list(arg.trim_start_matches("--waiting-when-tool="));
            }
            other => bail!("unknown argument {other:?}"),
        }
        i += 1;
    }
    Ok(out)
}

fn parse_tool_list(s: &str) -> Vec<String> {
    s.split(',')
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
        .collect()
}

/// Extract the `prompt` field from a Claude UserPromptSubmit payload.
pub fn read_prompt_from_bytes(data: &[u8]) -> Option<String> {
    if data.is_empty() {
        return None;
    }
    let v: serde_json::Value = serde_json::from_slice(data).ok()?;
    let p = v.get("prompt")?.as_str()?;
    (!p.is_empty()).then(|| p.to_string())
}

/// Extract `tool_name` from a Claude PreToolUse payload.
pub fn read_tool_name_from_bytes(data: &[u8]) -> Option<String> {
    if data.is_empty() {
        return None;
    }
    let v: serde_json::Value = serde_json::from_slice(data).ok()?;
    let t = v.get("tool_name")?.as_str()?.trim();
    (!t.is_empty()).then(|| t.to_string())
}

/// Run `report-status`: parse args (+stdin), resolve identity, and apply a
/// partial status write. Returns the workspace id that was written, or `None`
/// when it was a silent no-op (missing identity / unknown row / invalid state).
///
/// `viewing` is whether the user is currently attached to the session — passed
/// by the caller (false in v1; the badge-suppression query is a follow-up). A
/// bad state string is an error; everything else degrades to `Ok(None)`.
pub fn run(
    store: &Store,
    args: &ReportArgs,
    stdin: &mut dyn Read,
    ident: &Ident,
    viewing: bool,
) -> Result<Option<WorkspaceId>> {
    let mut state = args.state.trim().to_ascii_lowercase();
    if state.is_empty() {
        bail!("--state is required");
    }
    // Validate against the closed set of raw strings (not via `Status::parse`,
    // which folds unknown strings to Idle). Matches Go `validReportStates`.
    if !matches!(state.as_str(), "working" | "idle" | "waiting" | "exited") {
        bail!("invalid --state {state:?} (want working|idle|waiting|exited)");
    }

    let mut prompt = args.prompt.clone();
    // Read stdin once and route both extractions through it.
    if args.prompt_stdin || !args.waiting_tools.is_empty() {
        let mut buf = Vec::new();
        // Best-effort: a malformed payload never breaks the agent turn.
        let _ = stdin.read_to_end(&mut buf);
        if args.prompt_stdin {
            if let Some(p) = read_prompt_from_bytes(&buf) {
                prompt = Some(p);
            }
        }
        if !args.waiting_tools.is_empty() {
            if let Some(tool) = read_tool_name_from_bytes(&buf) {
                if args.waiting_tools.iter().any(|t| t == &tool) {
                    state = "waiting".to_string();
                }
            }
        }
    }
    let status = Status::parse(&state);
    let prompt = prompt.map(|p| p.trim().to_string());

    if ident.workspace.is_empty() {
        return Ok(None);
    }

    let Some((id, current_unread)) = resolve(store, ident)? else {
        return Ok(None);
    };

    // ActivePrompt lifecycle.
    let prompt_write: Option<String> = match prompt.as_deref() {
        Some(p) if !p.is_empty() => Some(p.to_string()),
        _ if status.is_exited() || status == Status::Idle => Some(String::new()),
        _ => None,
    };
    // Unread lifecycle.
    let unread = if status.is_exited() {
        false
    } else if status.wants_attention() {
        !viewing
    } else {
        current_unread
    };

    store.update_status(&id, status, prompt_write.as_deref(), unread)?;
    Ok(Some(id))
}

/// Resolve identity to a stored `(WorkspaceId, current_unread)`. Prefers an
/// exact repo-root match; falls back to a repo-basename lookup with the same
/// collision handling as Go (exactly-one match wins; ambiguity no-ops).
fn resolve(store: &Store, ident: &Ident) -> Result<Option<(WorkspaceId, bool)>> {
    if !ident.repo_root.is_empty() {
        let root = crate::repo::canonicalize_repo_root(&ident.repo_root);
        let id = WorkspaceId::new(root, ident.workspace.clone());
        return Ok(store.get_workspace(&id)?.map(|w| (id, w.unread)));
    }
    if ident.repo.is_empty() {
        return Ok(None);
    }
    // Basename fallback.
    let mut candidates: Vec<String> = store
        .repo_roots()?
        .into_iter()
        .filter(|root| basename(root) == ident.repo)
        .collect();
    candidates.sort();
    let mut matches: Vec<WorkspaceId> = Vec::new();
    for root in &candidates {
        let id = WorkspaceId::new(root.clone(), ident.workspace.clone());
        if store.get_workspace(&id)?.is_some() {
            matches.push(id);
        }
    }
    match matches.len() {
        0 => Ok(None), // Row not yet created — nothing to write.
        1 => {
            let id = matches.into_iter().next().unwrap();
            let unread = store.get_workspace(&id)?.map(|w| w.unread).unwrap_or(false);
            Ok(Some((id, unread)))
        }
        _ => Ok(None), // Ambiguous basename collision — drop rather than misroute.
    }
}

fn basename(path: &str) -> String {
    path.trim_end_matches('/')
        .rsplit('/')
        .next()
        .unwrap_or(path)
        .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use awp_core::Workspace;
    use std::io::Cursor;

    fn store_with(repo: &str, name: &str, status: Status, unread: bool, prompt: &str) -> Store {
        let store = Store::open_in_memory().unwrap();
        store
            .upsert_workspace(&Workspace {
                repo_root: repo.into(),
                name: name.into(),
                path: format!("{repo}/{name}"),
                status,
                unread,
                active_prompt: (!prompt.is_empty()).then(|| prompt.to_string()),
                ..Default::default()
            })
            .unwrap();
        store
    }

    fn ident(repo_root: &str, workspace: &str) -> Ident {
        Ident {
            workspace: workspace.into(),
            repo: String::new(),
            repo_root: repo_root.into(),
        }
    }

    #[test]
    fn waiting_sets_unread_and_writes_prompt() {
        let store = store_with("/r", "w", Status::Idle, false, "");
        let args = ReportArgs {
            state: "waiting".into(),
            prompt: Some("confirm?".into()),
            ..Default::default()
        };
        let id = run(
            &store,
            &args,
            &mut Cursor::new(b""),
            &ident("/r", "w"),
            false,
        )
        .unwrap()
        .unwrap();
        let w = store.get_workspace(&id).unwrap().unwrap();
        assert_eq!(w.status, Status::Waiting);
        assert!(w.unread);
        assert_eq!(w.active_prompt.as_deref(), Some("confirm?"));
    }

    #[test]
    fn working_preserves_existing_unread_and_prompt() {
        let store = store_with("/r", "w", Status::Waiting, true, "keep");
        let args = ReportArgs {
            state: "working".into(),
            ..Default::default()
        };
        let id = run(
            &store,
            &args,
            &mut Cursor::new(b""),
            &ident("/r", "w"),
            false,
        )
        .unwrap()
        .unwrap();
        let w = store.get_workspace(&id).unwrap().unwrap();
        assert_eq!(w.status, Status::Working);
        assert!(w.unread, "working preserves prior unread");
        assert_eq!(w.active_prompt.as_deref(), Some("keep"));
    }

    #[test]
    fn idle_clears_prompt() {
        let store = store_with("/r", "w", Status::Working, false, "midtask");
        let args = ReportArgs {
            state: "idle".into(),
            ..Default::default()
        };
        let id = run(
            &store,
            &args,
            &mut Cursor::new(b""),
            &ident("/r", "w"),
            false,
        )
        .unwrap()
        .unwrap();
        let w = store.get_workspace(&id).unwrap().unwrap();
        assert_eq!(w.active_prompt, None);
    }

    #[test]
    fn exited_clears_unread() {
        let store = store_with("/r", "w", Status::Waiting, true, "x");
        let args = ReportArgs {
            state: "exited".into(),
            ..Default::default()
        };
        let id = run(
            &store,
            &args,
            &mut Cursor::new(b""),
            &ident("/r", "w"),
            false,
        )
        .unwrap()
        .unwrap();
        let w = store.get_workspace(&id).unwrap().unwrap();
        assert!(!w.unread);
    }

    #[test]
    fn viewing_suppresses_unread() {
        let store = store_with("/r", "w", Status::Idle, false, "");
        let args = ReportArgs {
            state: "waiting".into(),
            ..Default::default()
        };
        let id = run(
            &store,
            &args,
            &mut Cursor::new(b""),
            &ident("/r", "w"),
            true,
        )
        .unwrap()
        .unwrap();
        assert!(!store.get_workspace(&id).unwrap().unwrap().unread);
    }

    #[test]
    fn prompt_stdin_extracts_from_payload() {
        let store = store_with("/r", "w", Status::Idle, false, "");
        let args = ReportArgs {
            state: "working".into(),
            prompt_stdin: true,
            ..Default::default()
        };
        let payload = br#"{"prompt":"do the thing"}"#;
        let id = run(
            &store,
            &args,
            &mut Cursor::new(payload),
            &ident("/r", "w"),
            false,
        )
        .unwrap()
        .unwrap();
        assert_eq!(
            store
                .get_workspace(&id)
                .unwrap()
                .unwrap()
                .active_prompt
                .as_deref(),
            Some("do the thing")
        );
    }

    #[test]
    fn waiting_when_tool_overrides_state() {
        let store = store_with("/r", "w", Status::Working, false, "");
        let args = ReportArgs {
            state: "working".into(),
            waiting_tools: vec!["AskUserQuestion".into()],
            ..Default::default()
        };
        let payload = br#"{"tool_name":"AskUserQuestion"}"#;
        let id = run(
            &store,
            &args,
            &mut Cursor::new(payload),
            &ident("/r", "w"),
            false,
        )
        .unwrap()
        .unwrap();
        assert_eq!(
            store.get_workspace(&id).unwrap().unwrap().status,
            Status::Waiting
        );
    }

    #[test]
    fn missing_identity_is_silent_noop() {
        let store = store_with("/r", "w", Status::Idle, false, "");
        let args = ReportArgs {
            state: "working".into(),
            ..Default::default()
        };
        let empty = Ident::default();
        assert!(run(&store, &args, &mut Cursor::new(b""), &empty, false)
            .unwrap()
            .is_none());
    }

    #[test]
    fn unknown_row_is_silent_noop() {
        let store = store_with("/r", "w", Status::Idle, false, "");
        let args = ReportArgs {
            state: "working".into(),
            ..Default::default()
        };
        assert!(run(
            &store,
            &args,
            &mut Cursor::new(b""),
            &ident("/r", "nope"),
            false
        )
        .unwrap()
        .is_none());
    }

    #[test]
    fn invalid_state_is_error() {
        let store = store_with("/r", "w", Status::Idle, false, "");
        let args = ReportArgs {
            state: "bananas".into(),
            ..Default::default()
        };
        assert!(run(
            &store,
            &args,
            &mut Cursor::new(b""),
            &ident("/r", "w"),
            false
        )
        .is_err());
    }

    #[test]
    fn basename_fallback_resolves_unique_match() {
        let store = Store::open_in_memory().unwrap();
        store
            .upsert_workspace(&Workspace {
                repo_root: "/home/me/proj".into(),
                name: "w".into(),
                path: "/home/me/proj/w".into(),
                ..Default::default()
            })
            .unwrap();
        let args = ReportArgs {
            state: "working".into(),
            ..Default::default()
        };
        let id_env = Ident {
            workspace: "w".into(),
            repo: "proj".into(),
            repo_root: String::new(),
        };
        let id = run(&store, &args, &mut Cursor::new(b""), &id_env, false)
            .unwrap()
            .unwrap();
        assert_eq!(id.repo_root, "/home/me/proj");
    }

    #[test]
    fn parse_args_rejects_unknown_flag() {
        assert!(parse_args(&["--bogus".to_string()]).is_err());
    }
}
